### PR TITLE
Alternative indenter

### PR DESCRIPTION
**Pull request by lefessan**
This indenter is much simpler than the previous one. It iterates on a list of tokens for the whole buffer, and uses a per-division automaton.

The current limitations are listed in the `main.ml` file. It has been tested on 1000+ programs from GnuCOBOL testsuite, on the NIST files and the X-COBOL testsuite, and the differences are usually a matter of personal choice rather than standard (or wrong initial indentation).

Also, it is completely independent from other packages from SuperBOL (it has its own lexer for example), and it will probably only use the per-project configuration to set indenting options.

Current configuration is done by `.superbol-indent` file in any parent directory, and the option `--gen-config` can be used to create such a file in the current directory.

For debugging, there is a `verbose` variable in `cobol_indent/engine.ml` that can be turned to `true`.

### CODE REVIEW

**Comment by lefessan** [comment]
* I think it is now ready for beta testing, as I am now pleased by the results on all examples I have tested on.
* Ideally, we should also provide an Emacs mode, maybe inspired by the one of `ocp-indent`
* There are two cases that we should think deeper about:
    * How to deal with tabs. Currently, the indenter only cares about a tab in the left margin, assumed of size 8, and all other tabs are seen as equivalent to spaces
    * How to deal with the right maring. The editor in the indenter correctly handles comments in the right margin, but what about vscode ?

**Comment by nberth** [comment]
Seems nice to me. I think you just need to `opam exec -- dune runtest test/lsp --auto-promote` to pass the CI checks.

For GNU/Emacs integration, maybe a standalone executable "superbol-indent` would be sufficient, as it would be much lighter than the whole LSP server.

File: test/lsp/lsp_formatting.ml [hunk 1]
```diff
@@ -100,17 +101,18 @@ let%expect_test "formatting-request-nested-if" =
...
+    PROCEDURE DIVISION.
+    para-1.
+        IF X>9
+        THEN
+           IF X>6
+           THEN
+              DISPLAY "2"
+           else
+              move 1 to x
+           else
```
Comment (lefessan) [inline test/lsp/lsp_formatting.ml:114]: This is a bug

**Review by nberth — APPROVED** [review]
Just some quick comments here and there.

(Still, half-wondering why you did not use `Cobol_preproc.fold_source_lines` like the old `cobol_indent` did. I think that'd have provided you with most of what you reconstructed, in addition with support for dynamic change of source formats.)

File: src/lsp/cobol_indent/scanner.ml [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+  (*
+  Printf.eprintf "FILE SIZE: %d\n%!" len;
+  Printf.eprintf "FORMAT: { free=%b, skip_before=%d, skip_after=%d }\n"
+    source_format.free source_format.skip_before source_format.skip_after;
+*)
+  let lines = ref [] in
+
+  let max_text_len =
+    if source_format.free then
+      65536
```
[the remark this answers is not in the document]
Comment (lefessan) [reply src/lsp/cobol_indent/scanner.ml:132]: I just needed to pick a "big enough" value, I could have chosen something much bigger... but not `maxint`, as I didn't want arithmetic operations to overflow.

File: src/lsp/cobol_indent/types.ml [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+type range = {
+  start_line :int ;
+  end_line   :int ;
+}
+
+type source_format = {
+  name : string ;
+  free : bool  ;      (* has column identifier *)
+  skip_before : int ;      (* fixed = 6 *)
+  max_text_length : int ;  (* fixed = 8..72 = 65 chars *)
```
[the remark this answers is not in the document]
Comment (lefessan) [reply src/lsp/cobol_indent/types.ml:31]: Yes, my current solution is to only support the formats that fit this description...

**Comment by lefessan** [comment]
> (Still, half-wondering why you did not use `Cobol_preproc.fold_source_lines` like the old `cobol_indent` did. I think that'd have provided you with most of what you reconstructed, in addition with support for dynamic change of source formats.)

Maybe, but (1) I didn't know about the existence of that function, and (2) dynamic changes of source formats would make some other parts of the indenter more complex (such as the patching engine), I prefer to wait to see it being useful in practice before taking hours to implement it correctly.

**Pull request merged**